Utility methods update

CptCrispyCrunchy opened the pull request:
-Added a method to center a template around its geometrical center in the XZ plan.

-Moved the `SpawnBlockRegionComponent.getBoundingBox()` to a new utility class to prepare for the engine update

skaldarnar commented on src/main/java/org/terasology/structureTemplates/util/BlockRegionToolbox.java:
```diff
@@ -0,0 +1,72 @@
+ * A collection of little helper to ease the work with structure templates.
+ **/
+
+public class BlockRegionToolbox {
```
I would name it `BlockRegionUtils`

CptCrispyCrunchy replied:
done :)

skaldarnar commented on src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java:
```diff
@@ -24,6 +24,7 @@
 import org.terasology.world.block.BlockUri;
 
 import static org.junit.Assert.assertEquals;
+import static org.terasology.structureTemplates.util.BlockRegionToolbox.getBoundingBox;
```
Try to avoid static imports.

CptCrispyCrunchy replied:
fixed :)

skaldarnar commented on src/main/java/org/terasology/structureTemplates/util/BlockRegionToolbox.java:
```diff
@@ -0,0 +1,72 @@
+
+public class BlockRegionToolbox {
+
+    public static BlockRegionMovement setOnCenterXZ(SpawnBlockRegionsComponent spawnBlockRegionsComponent) {
```
Probably add a general method to center along any (combination) of axis and provide convenience methods like this one.

CptCrispyCrunchy commented:
@flo : ping! DynamicCities would need that PR to run stable on the event tomorrow. Would be cool if you can take a look at it :)

flo commented on src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java:
```diff
@@ -0,0 +1,72 @@
+
+public class BlockRegionUtilities {
+
+    public static BlockRegionMovement setOnCenterXZ(SpawnBlockRegionsComponent spawnBlockRegionsComponent) {
```
I think it should not be named "set" as it is not a setter and does not set something.

Also there might be the need to determine the center for other reasons than the creation of BlockRegionMovement. So I would go for a method say like "Vector3i determineBottomCenterOf".

skaldarnar replied:
I still would go for an even more general version, like
```Java
public static Vector3i determineCenterPointAlongAxisOf(SpawnBlockRegionsComponent component, Axis... axis)
```
What do you think? `determineBottomCenterPointOf` could be implemented as:
```Java
public static Vector3i determineBottomCenterPointOf(SpawnBlockRegionsComponent component) {
    return determineCenterPointAlongAxisFor(component, Axis.X, Axis.Z);
}
```

P.S.: I don't think we have general axis descriptors yet, only [`Direction`](https://github.com/MovingBlocks/Terasology/blob/develop/engine/src/main/java/org/terasology/math/Direction.java).

flo replied:
I can't think of a scenario where you would need the general version of it. I would say there is more important stuff than a general version of it, since there would be also more to test.

skaldarnar replied:
Hm, maybe my thoughts were too complicated. I had some scenarios in mind were you would like to align a structure to something else, maybe placing it under the ceiling of a big cave, or aligning with neighbouring structures. But that can also be achieved with the simple version for bottomCenter and the general bounding box I guess...

CptCrispyCrunchy replied:
As for the naming: It sort of "sets" all blocks to be centered around 0,0. But you are right, it is probably a bit confusing like this.

Adding additional centering methods could be a thing for the future but as long as it doesn't needed I deem it unnecessary too. Though I guess it could be used for some cool clipping stuff later on. I would implement just additional methods though (would only be 2 more) instead of defining an axis class etc.

flo commented on src/test/java/org/terasology/structureTemplates/components/BlockRegionBoundingBoxTest.java:
```diff
@@ -46,9 +47,9 @@ public void testBoundingBox() {
-        assertEquals(bb.max(), new Vector3i(6, 1, 5));
-        assertEquals(bb.min(), new Vector3i(2, 0, -1));
+        Region3i bb = BlockRegionUtilities.getBoundingBox(blockRegionA);
+        Assert.assertEquals(bb.max(), new Vector3i(6, 1, 5));
```
Was this change intended. I prefer usually assertEquals without the Assert infront.

skaldarnar replied:
oh, that accounts probably accounts to me advising to avoid static imports... I guess static imports are fine for `Assert` in unit tests, but `BlockRegionUtils` should not be statically imported imho.

flo replied:
yes, especially getBoudningBox should not be statically imported as it could be confused with a local getter.

flo commented on src/main/java/org/terasology/structureTemplates/util/BlockRegionUtilities.java:
```diff
@@ -0,0 +1,72 @@
+/**
+ * A collection of little helper to ease the work with structure templates.
+ **/
+
```
Is the empty line here intended?

It is however good that you added a javadoc. However I would suggest that you be more specific as the whole  module is about structure templates.

CptCrispyCrunchy replied:
Oh the empty line was by accident. 
As for being more specific: I guess I adding some method descriptions instead of that line could be more helpful then.

flo commented on src/main/java/org/terasology/structureTemplates/components/SpawnBlockRegionsComponent.java:
```diff
@@ -31,35 +31,6 @@
 public class SpawnBlockRegionsComponent implements Component {
     public List<RegionToFill> regionsToFill = new ArrayList<>();
 
-    public Region3i getBoundingBox() {
```
thanks for moving it to a utility  class, that is much better

CptCrispyCrunchy commented:
@flo latest commit should have everything right now. Refactored method name, added comments to the methods instead of the class, even removed empty lines :D and `determineBottomCenter()` returns a Vector3i now.

flo commented:
merged
